### PR TITLE
Speed up bigtiff test using `np.empty` and smaller data

### LINKED ISSUE #7382: [Maint, CI] Rethink napari/utils/_tests/test_io.py::test_imsave_large_file -- it's very slow
## 🧰 Task

On CI test runs, `napari/utils/_tests/test_io.py::test_imsave_large_file` takes >1 min and on some runs >2 min, making it ~10-15% of total run time.

Currently the test is writing and reading a 4Gb tiff:
https://github.com/napari/napari/blob/b457880f2b8b33b3b966f4a8dcb074dbf87b04d0/napari/utils/_tests/test_io.py#L75-L93

I think we can trust that if `bigtiff=True` is passed to `tifffile.imwrite` that it really works...
maybe we leave the existing test and skip it on CI and add a test with a mock for CI?

### PR DESCRIPTION
**Pull request by Czaki**
# References and relevant issues

closes #7382

# Description

We don't actually need big data because we monkeypatch tifffile's writer to raise the bigtiff data error even on small arrays. Additionally, we can use np.empty rather than np.random.random because we don't care about the array contents.

### CODE REVIEW

**Comment by psobolewskiPhD** [comment]
Wow, huge differences on the 3.12 initial:
`  4.68s call     napari/utils/_tests/test_io.py::test_imsave_large_file`
versus:
`58.24s call     napari/utils/_tests/test_io.py::test_imsave_large_file`
without this.
On 3.9 initial similar:
`4.59s call     napari/utils/_tests/test_io.py::test_imsave_large_file`
versus 
`58.60s call     napari/utils/_tests/test_io.py::test_imsave_large_file`

**Review by psobolewskiPhD — APPROVED** [review]

**Comment by psobolewskiPhD** [comment]
I think this is more than enough to consider the issue closed.

**Comment by Czaki** [comment]
@psobolewskiPhD Just pushed change for more speedup. It is bellow 1s now :) 

@willingc You may see from the above @psobolewskiPhD comment, that using random data in tests may impact, not only in random crash of data, but also in slowdown. 

Unfortunately, I do not convince other core devs to not accept using random data in tests. 
I also forgot to open followup PR to #7242 to speedup test.

**Comment by willingc** [comment]
> @willingc You may see from the above @psobolewskiPhD comment, that using random data in tests may impact, not only in random crash of data, but also in slowdown.

> Unfortunately, I do not convince other core devs to not accept using random data in tests.

Hmm... I think that there is a case to be made for your approach as well @Czaki. Perhaps add as a separate test that we run on only one OS version/python/qt backend config.

Nice work getting this sped up for now @Czaki.

Tests will be key for the work to refactor and pay down tech debt.

**Review by willingc — APPROVED** [review]

File: napari/utils/_tests/test_io.py
Comment (jni) [inline napari/utils/_tests/test_io.py]:
```suggestion
    # create image data. It can be <4GB compressed because we
    # monkeypatch tifffile.imwrite to raise an error if bigtiff is not set
```

**Comment by jni** [comment]
> Unfortunately, I do not convince other core devs to not accept using random data in tests.

I don't recall any attempt to convince me and I apologise if I blocked anything! 😅 random is just the default my brain goes to.

**Comment by Czaki** [comment]
> I don't recall any attempt to convince me and I apologise if I blocked anything! 😅 random is just the default my brain goes to.

Not blocked. Just create and accept PR that contains it. 

Yes, I know that it require to change mindset.

File: napari/utils/_tests/test_io.py
Comment (psobolewskiPhD) [inline napari/utils/_tests/test_io.py]:
```suggestion
    # monkeypatch tifffile.imwrite to raise a struct.error if bigtiff is not set
    # this checks the try/except in io.imsave_tiff resulting in bigtiff being set 
    data = np.empty((20, 200, 200), dtype='uint16')
```

As is, I think the test is rather counter intuitive. Maybe this extra comment helps point people to the right code to make it make sense?

Comment (psobolewskiPhD) [reply napari/utils/_tests/test_io.py]: At least for me the new version (small file) made no sense and I expected it to fail until I followed the chain from `imsave` to `imsave_tiff`

Comment (Czaki) [reply napari/utils/_tests/test_io.py]: I know, but without this patch, the compression already keeps the size of the image below 4GB, so it was not the big tiff. 

So this is part that require good clarification, but I do not feel to be a good person, to validate if it is.

Comment (psobolewskiPhD) [reply napari/utils/_tests/test_io.py]: I added a line to the comment. 🤷

Comment (Czaki) [reply napari/utils/_tests/test_io.py]: @jni Could you help to check if it is enough?

Comment (willingc) [reply napari/utils/_tests/test_io.py]: I suggest adding a docstring at the top of the test  to reduce cognitive load and make clear what the test's setup and behavior should be.

```python
    """Test saving a big tiff file with a size of 4GB.
    
    Set up
    1. Create a 4G image file of random data.
    2. Ensure that the image file is set to bigtiff.
       We check that it is set to bigtiff since compression may reduce the file size below 4GB.

    Test that imsave saves the bigtiff file correctly.
    """
```

The additional context helps the line comment make sense.

**Review by willingc — COMMENTED** [review]
Added a suggestion for a docstring.

File: napari/utils/_tests/test_io.py
Comment (Czaki) [reply napari/utils/_tests/test_io.py]: Ok. Maybe I should provide more details. 

In general, the TIFF file format specification predicts only 4GB space for data. However, it not, not mean that all array of size above 4GB need to be saved using bigtiff extensions. As we use compression when saving tiff, the bigtiff mode is required, only if the data size **after compression** is above 4GB. 

So when the test was developed, it happened that random-generated data could be compressed to fit in 4GB. 

So we added monkeypatching, to always enforce bigtiff mode in this test. 

So with this PR, we create only small data, but, because of monkeypatch, it is saved as bigtiff. 

The bigtiff is an extension of tiff and could be used with even small files. However, there are still some apps that do not support it. Therefore, we decided to use it only for files that do not fit in 4GB space limit. 

We do not want to fight with generation of real data, that will be really slow. And may create memory problems.

Comment (willingc) [reply napari/utils/_tests/test_io.py]: Let's see if I can rephrase what you said.

    """Test saving a bigtiff file.

    In our application, we use compression when saving tiff. The bigtiff mode is required, only if the
    data size after compression is above 4GB.
    
    Set up
    1. Create a randomly-generated data file to represent an image. 
    2. Monkeypatch the data file to enforce bigtiff mode. This allows us to
       use a smaller data file to simulate a bigtiff while keeping the test run fast.

    Test that imsave saves the file as bigtiff correctly.
    """

Comment (willingc) [reply napari/utils/_tests/test_io.py]: Thanks @Czaki. I will leave it up to you to add the docstring to the PR. Thanks!

**Comment by jni** [comment]
I pushed a slightly more elaborate version of @willingc's docstring suggestion. This is ready to merge afaic.

**Review by jni — APPROVED** [review]

**Review by willingc — APPROVED** [review]
Love this. Thanks @Czaki! 🚢  (and @jni :smile:)

**Pull request merged**
